### PR TITLE
[SE-0267] Improvements (Ep. 1)

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
cc @slavapestov 

- [x] Move diagnosing the where clause to Sema
- [x] Add thunk emission tests to `test/SILGen/vtable_thunks_reabstraction.swift`
- [x] Add type reconstruction tests to `test/TypeDecoder/`

### CODE REVIEW

File: test/TypeDecoder/nominal_types.swift [hunk 1]
```diff
@@ -153,6 +159,8 @@ extension STSContainer℠ where T == STSOuter {
 
 // DEMANGLE-TYPE: $s13nominal_types12STSContainerO8SubclassC11ExtraNestedCyAA8STSOuterV_Si_G
 // CHECK-TYPE: STSContainer<STSOuter>.Subclass<Int>.ExtraNested
+// DEMANGLE-TYPE: $s13nominal_types12STSContainerO8SubclassC16ExtraNestedWhereCyAA8STSOuterV_AEyAI_AIG_G
+// CHECK-TYPE: STSContainer<STSOuter>.Subclass<STSContainer<STSOuter>.Subclass<STSOuter>>.ExtraNestedWhere
```
Comment (AnthonyLatsis) [inline test/TypeDecoder/nominal_types.swift]: I would appreciate some advice on the proper way or utility to generate mangled names for types and declarations — doing this via `-emit-ir` and intuitive parsing is way too cumbersome to be the truth, especially for aliases that get stripped during silgen.

File: test/SILGen/vtable_thunks_reabstraction.swift [hunk 2]
```diff
@@ -557,3 +557,33 @@ class MoreGenericSub2<TT, T> : GenericBase<T> {
...
+}
+class SE_0267_Derived1<T>: SE_0267_Base1<T> {
+  override func foo() {}
+}
+
+// CHECK-LABEL: sil_vtable SE_0267_Derived1 {
+// CHECK-NEXT:   #SE_0267_Base1.foo!1: <T where T : SE_0267_P1> (SE_0267_Base1<T>) -> () -> () : @$s27vtable_thunks_reabstraction16SE_0267_Derived1C3fooyyFAA0D11_0267_Base1CADyyAA0D8_0267_P1RzlFTV [override]
+// CHECK-NEXT:   #SE_0267_Base1.init
+// CHECK-NEXT:   #SE_0267_Derived1.foo!1: <T> (SE_0267_Derived1<T>) -> () -> () : @$s27vtable_thunks_reabstraction16SE_0267_Derived1C3fooyyF
+// CHECK-NEXT:   #SE_0267_Derived1.deinit
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply test/SILGen/vtable_thunks_reabstraction.swift]: SE_0267_Base1.foo() can only be called where T conforms to `SE_0267_P1`, so the calling convention takes a witness table for the conformance.

SE_0267_Derived1.foo() can be called with a T that does not conform to `SE_0267_P1`, so we have to introduce a second vtable entry that does not take the witness table.

In your second example, the override can be massaged to have a calling convention compatible with the base method (we just have to return the `Int` value indirectly), so it doesn't need a new vtable entry.

File: test/TypeDecoder/nominal_types.swift (hunk 1 above)
Comment (slavapestov) [reply test/TypeDecoder/nominal_types.swift]: I don't have a good solution. When adding new tests here I just build with `-emit-ir -g` and grep through the resulting debug info.

**Review by slavapestov — APPROVED** [review]

File: test/SILGen/vtable_thunks_reabstraction.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnthonyLatsis) [reply test/SILGen/vtable_thunks_reabstraction.swift]: Thank you for breaking this up for me!

**Comment by AnthonyLatsis** [comment]
@slavapestov In case you forgot, I can't trigger the tests myself 😅

**Comment by xwu** [comment]
@swift-ci test

**Comment by AnthonyLatsis** [comment]
These failures are not related though :(

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@AnthonyLatsis Please go ahead and request commit access so you can trigger tests. You've contributed more than enough at this point.

**Comment by AnthonyLatsis** [comment]
@swift-ci Please test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test

**Comment by AnthonyLatsis** [comment]
Ah, this time it's because of #30499

**Comment by AnthonyLatsis** [comment]
@swift-ci please test

**Comment by AnthonyLatsis** [comment]
🍄

**Pull request merged**